### ISSUE TITLE
Add an exception mapper to gRPC services

### ISSUE THREAD

**Issue by wheeOs** [quarkusio/quarkus#28609]
### Description

I couldn't find a way to intercept (to be more precise) an exception response which ends up as an response with `UNKNOWN `status. In my case I would like to map specific exceptions to specific status other than UNKNOWN.
Here is an exception I get, which I would like to be considered as `UNAUTHENTICATED`, but it ends up as `UNKNOWN `unfortunately:

```
2022-10-14 18:01:21,371 WARN  [io.qua.grp.stu.ServerCalls] (vert.x-eventloop-thread-1) gRPC service threw an exception other than StatusRuntimeException: io.quarkus.security.AuthenticationFailedException
	at io.quarkus.oidc.runtime.OidcIdentityProvider.verifyTokenUni(OidcIdentityProvider.java:289)
	at io.quarkus.oidc.runtime.OidcIdentityProvider.createSecurityIdentityWithOidcServer(OidcIdentityProvider.java:151)
	at io.quarkus.oidc.runtime.OidcIdentityProvider$3.apply(OidcIdentityProvider.java:134)
	at io.quarkus.oidc.runtime.OidcIdentityProvider$3.apply(OidcIdentityProvider.java:128)
```

### Implementation ideas

Maybe allow `ServerInterceptor`s to also intercept responses / errors?

**Comment by cescoffier** [comment]
Can't a ServerInterceptor do this already?

**Comment by wheeOs** [comment]
@cescoffier the only potential place where I could catch the Exception would be in `ServerCall.Listener<ReqT>.onHalfClose`. But:

```
            override fun onHalfClose() {
                try {
                    super.onHalfClose()
                } catch (ex: Throwable) {
                    println(ex)
                }
            }
```
my catch clause gets never reached because the exception is caught earlier in `ServerCalls.java` and the response is passed to a Mutiny downstream which we don't have access to from a `ServerInterceptor`. Hence, wouldn't it make sense to also being able  to intercept the resulting response?

**Comment by cescoffier** [comment]
I see. 
Unfortunately, there is not much we can do, except, maybe, pass the exception into a gRPC context. We need to call onError to pass the wrapped error to the gRPC downstream. If we don't, it may hang forever. 

We could imagine the second level of interception - something like `GrpcServerFailureInterceptor,` which would be called in the `catch` block with the original exception would have the possibility to process the failure synchronously. However, I would rather to that with Fault Tolerance and as an @Retry or @Fallback on the service method.

**Comment by geoand** [comment]
I don't know anything about gRPC, but this request sounds a lot like the `ExceptionMapper` from JAX-RS

**Comment by cescoffier** [comment]
yes, it's relatively similar. gRPC interceptors are rather low-level.

**Comment by cescoffier** [comment]
@alesj do you think we can implement a exception mapper in the new gRPC server support?

**Comment by alesj** [comment]
>> do you think we can implement a exception mapper in the new gRPC server support?

@cescoffier I guess we could add some exception mapper here:
* https://github.com/quarkusio/quarkus/blob/main/extensions/grpc/runtime/src/main/java/io/quarkus/grpc/runtime/GrpcServerRecorder.java#L172

Or if @vietj has some other idea?

**Comment by alesj** [comment]
Linking the issue: #29328

**Comment by cescoffier** [comment]
Fixed by #29419
